Extended language service

danieltre23 opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
Previous commit added the extended template diagnostics integration with the compiler, but the language service uses a different function for non-ts files. Currently the language service would not generate extended diagnostics for the non-ts files.

Issue Number: #42966 


## What is the new behavior?
The language service can now generate extended template diagnostics for non-ts files.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

## Other information
Ignore first 5 commit, since they are in a separate PR #43107

danieltre23 committed:
refactor(compiler-cli): use `TemplateDiagnostic` in `extended` package

Return `TemplateDiagnostic` instead of `ts.Diagnostic` when getting the
extended template diagnostics. This makes the integration with the
language service easier. This also fixes the error code and now uses the
`ngErrorCode` for extended template diagnostics.

Refs #42966

danieltre23 committed:
refactor(compiler-cli): integrate extended diagnostics with the language service

Add the call to get the extended template diagnostics in
the compiler's `getDiagnosticsForComponent`. This makes showing
extended diagnostics in non-ts files posible.

Refs #42966

danieltre23 committed:
test(language-service): add test for extended diagnostics

Add a test to make sure extended template diagnostics are being correctly
generated in the language service for non-ts files.

Refs #42966

dgp1130 commented on packages/compiler-cli/src/ngtsc/typecheck/testing/index.ts:
```diff
@@ -20,9 +20,8 @@ import {ClassDeclaration, isNamedClassDeclaration, TypeScriptReflectionHost} fro
-import {ProgramTypeCheckAdapter, TemplateTypeChecker, TypeCheckContext} from '../api';
+import {ProgramTypeCheckAdapter, TemplateDiagnostic, TemplateTypeChecker, TypeCheckContext} from '../api';
 import {TemplateId, TemplateSourceMapping, TypeCheckableDirectiveMeta, TypeCheckBlockMetadata, TypeCheckingConfig} from '../api/api';
-import {TemplateDiagnostic} from '../diagnostics';
```
Nit: I think we can also remove this dependency from the associated build file. It's good to do that to keep individual targets as small as possible where we can.

dgp1130 reviewed (approved):
Looks great! I don't really see any issues at all with this. It's particularly helpful that the first commit is just a pure move and makes this much easier to review as a result.

atscott reviewed (changes requested):
Could you also add a test for inline templates? Otherwise this change LGTM

danieltre23 commented:
> Could you also add a test for inline templates? Otherwise this change LGTM

I actually added that test in #43107 because it worked with the compiler integration. But I can move it here if you'd prefer that.

https://github.com/angular/angular/pull/43107/files#:~:text=%7D)%3B-,it(%27should%20produce%20invalid%20banana%20in%20box%20warning%27%2C%20()%20%3D%3E%20%7B,-const%20files%20%3D%20%7B

atscott commented:
Oh, I see. That's alright, as long as we have coverage somewhere for it.

dgp1130 commented on packages/compiler-cli/src/ngtsc/typecheck/src/checker.ts:
```diff
@@ -245,6 +247,14 @@ export class TemplateTypeCheckerImpl implements TemplateTypeChecker {
           diag => convertDiagnostic(diag, fileRecord.sourceManager)));
       diagnostics.push(...shimRecord.genesisDiagnostics);
 
+      if (extendedTemplateDiagnsotics === true) {
```
Nit: `if (extendedTemplateDiagnostics) { /* ... */ }`.

danieltre23 commented:
Addressed some comments from #42984, in particular added the `__ngCode` brand that Alex suggested, PTAL.

danieltre23 committed:
refactor(compiler-cli): move `ExtendedTemplateChecker` to `LazyCompilationState`

Create the `ExtendedTemplateChecker` during `makeCompilation`, that way
it's avaiblable in the whole compiler.

Refs #42966

danieltre23 committed:
refactor(compiler-cli): move `getDiagnosticsForComponent` to compiler

Change the language service to call `getDiagnosticsForComponent` from
the compiler instead of the `TemplateTypeChecker`.

Refs #42966

danieltre23 committed:
refactor(compiler-cli): rename `ExtendedTemplateChecker` get diagnostics fucntion

Rename `getExtendedTemplateDiagnosticsForComponent` to
`getDiagnosticsForComponent` since it's implied they are extended
diagnostics.

Refs #42966
